### PR TITLE
ARROW-16464: [C++][CI][GPU] Add CUDA CI

### PR DESCRIPTION
**Pull request by assignUser**

### CODE REVIEW

**Comment by assignUser** [comment]
@github-actions crossbow submit test-cuda-*

**Comment by pitrou** [comment]
It's weird that the compilation command line for Arrow CUDA (and only Arrow CUDA) did not include `-isystem /arrow/cpp/thirdparty/flatbuffers/include` as expected. @kou Do you have an idea why that would be?

**Comment by pitrou** [comment]
More exactly, this is the compilation command that succeeded locally:
```
/usr/local/bin/sccache /usr/lib/ccache/c++  -DARROW_EXPORTING -DARROW_HAVE_RUNTIME_AVX2 -DARROW_HAVE_RUNTIME_AVX512 -DARROW_HAVE_RUNTIME_BMI2 -DARROW_HAVE_RUNTIME_SSE4_2 -DARROW_HAVE_SSE4_2 -DARROW_NO_DEPRECATED_API -DARROW_WITH_BROTLI -DARROW_WITH_BZ2 -DARROW_WITH_LZ4 -DARROW_WITH_SNAPPY -DARROW_WITH_ZLIB -DARROW_WITH_ZSTD -DBOOST_ALL_NO_LIB -DURI_STATIC_BUILD -Isrc -I/arrow/cpp/src -I/arrow/cpp/src/generated -isystem /usr/local/cuda/include -isystem /arrow/cpp/thirdparty/flatbuffers/include -isystem /arrow/cpp/thirdparty/hadoop/include -isystem xsimd_ep/src/xsimd_ep-install/include -Wno-noexcept-type  -fdiagnostics-color=always -ggdb -O0  -Wall -Wno-conversion -Wno-sign-conversion -Wunused-result -Werror -fno-semantic-interposition -msse4.2  -g -fPIC   -pthread -std=c++17 -MD -MT src/arrow/gpu/CMakeFiles/arrow_cuda_objlib.dir/cuda_arrow_ipc.cc.o -MF src/arrow/gpu/CMakeFiles/arrow_cuda_objlib.dir/cuda_arrow_ipc.cc.o.d -o src/arrow/gpu/CMakeFiles/arrow_cuda_objlib.dir/cuda_arrow_ipc.cc.o -c /arrow/cpp/src/arrow/gpu/cuda_arrow_ipc.cc
```

This is the line that failed on CI:
```
/usr/local/bin/sccache /usr/lib/ccache/c++  -DARROW_EXPORTING -DARROW_HAVE_RUNTIME_AVX2 -DARROW_HAVE_RUNTIME_AVX512 -DARROW_HAVE_RUNTIME_BMI2 -DARROW_HAVE_RUNTIME_SSE4_2 -DARROW_HAVE_SSE4_2 -DARROW_NO_DEPRECATED_API -DARROW_WITH_BROTLI -DARROW_WITH_BZ2 -DARROW_WITH_LZ4 -DARROW_WITH_SNAPPY -DARROW_WITH_ZLIB -DARROW_WITH_ZSTD -DURI_STATIC_BUILD -Isrc -I/arrow/cpp/src -I/arrow/cpp/src/generated -isystem /usr/local/cuda/include -Wno-noexcept-type  -fdiagnostics-color=always -ggdb -O0  -Wall -Wno-conversion -Wno-sign-conversion -Wunused-result -Werror -fno-semantic-interposition -msse4.2  -g -fPIC   -std=c++17 -MD -MT src/arrow/gpu/CMakeFiles/arrow_cuda_objlib.dir/cuda_arrow_ipc.cc.o -MF src/arrow/gpu/CMakeFiles/arrow_cuda_objlib.dir/cuda_arrow_ipc.cc.o.d -o src/arrow/gpu/CMakeFiles/arrow_cuda_objlib.dir/cuda_arrow_ipc.cc.o -c /arrow/cpp/src/arrow/gpu/cuda_arrow_ipc.cc
```

**Comment by assignUser** [comment]
The main difference between the crossbow job and my setup is that the cuda runner is arm64 could this play a role?

**Comment by pitrou** [comment]
And here is the expanded diff between both command lines:
```diff
--- good-cuda-cpp.txt	2022-10-25 15:45:21.569667959 +0200
+++ bad-cuda-cpp.txt	2022-10-25 15:45:34.889517034 +0200
@@ -13,19 +13,12 @@
 -DARROW_WITH_SNAPPY
 -DARROW_WITH_ZLIB
 -DARROW_WITH_ZSTD
--DBOOST_ALL_NO_LIB
 -DURI_STATIC_BUILD
 -Isrc
 -I/arrow/cpp/src
 -I/arrow/cpp/src/generated
 -isystem
 /usr/local/cuda/include
--isystem
-/arrow/cpp/thirdparty/flatbuffers/include
--isystem
-/arrow/cpp/thirdparty/hadoop/include
--isystem
-xsimd_ep/src/xsimd_ep-install/include
 -Wno-noexcept-type
 -fdiagnostics-color=always
 -ggdb
@@ -39,7 +32,6 @@
 -msse4.2
 -g
 -fPIC
--pthread
 -std=c++17
 -MD
 -MT
```

Besides the missing includes, we also see that `-pthread` and `-DBOOST_ALL_NO_LIB` have disappeared, which is weird.

**Comment by pitrou** [comment]
> The main difference between the crossbow job and my setup is that the cuda runner is arm64 could this play a role?

Is it? It looks like you're building/running a x86-64 image anyway, which is going to be slow (emulated).

**Comment by assignUser** [comment]
It is not, I was mixing things up as we also have a new arm runner, sorry for the confusion...

File: dev/tasks/docker-tests/github.cuda.yml [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+      {{ macros.github_checkout_arrow(fetch_depth=fetch_depth|default(1))|indent }}
+      # python 3.8 is installed on the runner, no need to install
+      - name: Install archery
+        run: python -m pip install -e arrow/dev/archery[docker]
+      - name: Execute Docker Build
+        shell: bash
+        env:
+        {{ macros.github_set_sccache_envvars()|indent(8) }}
+        run: |
+          archery docker run \
```
Comment (pitrou) [inline dev/tasks/docker-tests/github.cuda.yml:38]: So that we can see what the executed docker commands look like
```suggestion
          archery docker --debug run \
```

Comment (assignUser) [reply dev/tasks/docker-tests/github.cuda.yml:38]: I'll add it manualy to the workflow so we don't have to wait

Comment (assignUser) [reply dev/tasks/docker-tests/github.cuda.yml:38]: https://github.com/ursacomputing/crossbow/actions/runs/3321733516/jobs/5489783547

Comment (assignUser) [reply dev/tasks/docker-tests/github.cuda.yml:38]: https://github.com/ursacomputing/crossbow/actions/runs/3321733516/jobs/5489783547#step:5:18

Comment (assignUser) [reply dev/tasks/docker-tests/github.cuda.yml:38]: It looks like it is running it via docker cli instead of docker-compose: 
```
DEBUG:archery:Executing `['docker', 'run', '--rm', '-e', 'SETUPTOOLS_SCM_PRETEND_VERSION=10.0.0.dev509', '--gpus', 'all', '--shm-size', '2G', '-e', 'ARROW_BUILD_UTILITIES=OFF', '-e', 'ARROW_COMPUTE=OFF', '-e', 'ARROW_CSV=OFF', '-e', 'ARROW_CUDA=ON', '-e', 'ARROW_DATASET=OFF', '-e', 'ARROW_ENABLE_TIMING_TESTS=OFF', '-e', 'ARROW_FILESYSTEM=OFF', '-e', 'ARROW_GANDIVA=OFF', '-e', 'ARROW_GCS=OFF', '-e', 'ARROW_HDFS=OFF', '-e', 'ARROW_JEMALLOC=OFF', '-e', 'ARROW_JSON=OFF', '-e', 'ARROW_ORC=OFF', '-e', 'ARROW_PARQUET=OFF', '-e', 'ARROW_PLASMA=OFF', '-e', 'ARROW_S3=OFF', '-e', 'ARROW_SUBSTRAIT=OFF', '-e', 'ARROW_WITH_OPENTELEMETRY=OFF', '-e', 'AWS_ACCESS_KEY_ID=None', '-e', 'AWS_SECRET_ACCESS_KEY=None', '-e', 'CCACHE_COMPILERCHECK=content', '-e', 'CCACHE_COMPRESS=1', '-e', 'CCACHE_COMPRESSLEVEL=6', '-e', 'CCACHE_DIR=/ccache', '-e', 'CCACHE_MAXSIZE=1G', '-e', 'SCCACHE_BUCKET=None', '-e', 'SCCACHE_S3_KEY_PREFIX=sccache', '-v', '/runner/_work/crossbow/crossbow/arrow:/arrow:delegated', '-v', 'ubuntu-ccache:/ccache:delegated', 'apache/arrow-dev:amd64-ubuntu-20.04-cuda-11.0.3-cpp', '/bin/bash', '-c', '\n  /arrow/ci/scripts/cpp_build.sh /arrow /build &&\n  /arrow/ci/scripts/integration_skyhook.sh /build &&\n  /arrow/ci/scripts/cpp_test.sh /arrow /build']`
```

Comment (assignUser) [reply dev/tasks/docker-tests/github.cuda.yml:38]: same error when running with docker-compose directly https://github.com/ursacomputing/crossbow/actions/runs/3321875304/jobs/5490121964#step:5:2730

Comment (pitrou) [reply dev/tasks/docker-tests/github.cuda.yml:38]: Can you reproduce directly by ssh'ing into the runner?

**Comment by assignUser** [comment]
@kou could this change be the cause for the missing flags? https://github.com/apache/arrow/pull/12861/files#diff-cdd010da5f0f366c5098ad358c2f3479f4bb653b5871e639ed5333059b00493b

**Comment by assignUser** [comment]
The issue was caused by a mismatch between the docker images used for the runner and the job, they apparently have to be the same image or at least have the same CUDA version down to patch level.

**Comment by assignUser** [comment]
@github-actions crossbow submit test-cuda-*

**Comment by assignUser** [comment]
@raulcd could you have a look?

File: .env [hunk 2]
```diff
@@ -54,7 +54,7 @@ UBUNTU=20.04
 
 # Default versions for various dependencies
 CLANG_TOOLS=14
-CUDA=11.0.3
+CUDA=11.4.2
```
Comment (pitrou) [inline .env:57]: We can perhaps revert this change now?

**Comment by assignUser** [comment]
I rebased and edited the workflow to use the most recent commit as this issue was fixed by https://github.com/apache/arrow/pull/14501

File: .env (hunk 2 above)
Comment (assignUser) [reply .env:57]: As the cuda version is hardcoded we can change it

**Comment by assignUser** [comment]
The badge points to the crossbow workflow on the branch but the link points to the actual workflow. So when I edited the workflow file and triggered a new build the badge updates to working but the link still points to the old, failing build :D

**Review by pitrou — APPROVED** [review]
Thanks a lot for driving this @assignUser .
